### PR TITLE
Fixes ﻿#838 Rename and cloned rename working

### PR DESCRIPTION
**Pull request by rwmcintosh**
For clone rename or just rename, we are using this new dialog for expression profiles. You don't rename an expression profiles, instead you have to change the individual fields that comprise the name

image

As you type, the readonly 'Name' is updated in real time, as is the validation indicator.

image

### CODE REVIEW

File: src/MoBi.Presentation/Tasks/Interaction/ObjectBaseTask.cs [hunk 1]
```diff
@@ -1,128 +0,0 @@
-﻿using System.Collections.Generic;
```
Comment (rwmcintosh) [inline src/MoBi.Presentation/Tasks/Interaction/ObjectBaseTask.cs:1]: A lot of the changed files here are because I removed this task. It mostly renamed objects and up to now, all the renaming was completely common for all building blocks.

I needed to rename building blocks now in a type-sensitive way because the renaming presenter for Expression Profile would be different from most other cases.

Ultimately, this code was added to one of the base 'Interaction' tasks rather than delegated to this type-insensitive class.

There were also a couple of dependencies that needed to be injected as a result.

File: src/MoBi.Presentation/Tasks/Edit/EditTaskFor.cs [hunk 2]
```diff
@@ -66,7 +75,89 @@ public virtual void Edit(T objectToEdit)
       public virtual void Rename(T objectBase, IEnumerable<IObjectBase> existingObjectsInParent, IBuildingBlock buildingBlock)
       {
          var forbiddenNames = GetForbiddenNames(objectBase, existingObjectsInParent);
-         _context.AddToHistory(_interactionTask.Rename(objectBase, forbiddenNames, buildingBlock));
+         _context.AddToHistory(rename(objectBase, forbiddenNames, buildingBlock));
```
Comment (rwmcintosh) [inline src/MoBi.Presentation/Tasks/Edit/EditTaskFor.cs:78]: Here now instead of calling _interactionTask, which calls ObjectBaseTask, we can just do all the renaming here because this is a generic class.

File: src/MoBi.Presentation/Tasks/Interaction/InteractionTask.cs [hunk 3]
```diff
@@ -76,11 +73,6 @@ public class InteractionTask : IInteractionTask
          _serializationTask.SaveModelPart(entityToSerialize, fileName);
       }
 
-      public IMoBiCommand Rename<T>(T objectBase, IEnumerable<string> forbiddenNames, IBuildingBlock buildingBlock) where T : IObjectBase
```
Comment (rwmcintosh) [inline src/MoBi.Presentation/Tasks/Interaction/InteractionTask.cs:79]: Previously, InteractionTask.Rename just called ObjectBaseTask.Rename

File: src/MoBi.Core/Commands/RenameExpressionProfileBuildingBlockCommand.cs [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+      public override void RestoreExecutionData(IMoBiContext context)
+      {
+         base.RestoreExecutionData(context);
+         _expressionProfileBuildingBlock = context.Get<ExpressionProfileBuildingBlock>(ObjectId);
+      }
+
+      protected override void RenameBuildingBlock(IMoBiContext context)
+      {
+         base.RenameBuildingBlock(context);
+         var objectPathFactory = context.ObjectPathFactory;
```
Comment (msevestre) [inline src/MoBi.Core/Commands/RenameExpressionProfileBuildingBlockCommand.cs:39]: I'd suggest to add a comment here4 as this is not quite obvious why you need to mess around with path

File: src/MoBi.Core/Commands/RenameExpressionProfileBuildingBlockCommand.cs [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+         _expressionProfileBuildingBlock.Each(x =>
+         {
+            var objectPath = x.Path;
+            objectPath.Replace(_oldMoleculeName, _expressionProfileBuildingBlock.MoleculeName);
+            x.Path = objectPathFactory.CreateObjectPathFrom(objectPath);
```
Comment (msevestre) [inline src/MoBi.Core/Commands/RenameExpressionProfileBuildingBlockCommand.cs:44]: yeah this 3 lines of code feel really weird. You replace in the reference (same object) and then create a new entry ?

File: src/MoBi.Core/Commands/RenameObjectBaseCommand.cs [hunk 6]
```diff
@@ -39,15 +39,20 @@ protected override void ClearReferences()
       protected override void ExecuteWith(IMoBiContext context)
       {
          base.ExecuteWith(context);
+         RenameBuildingBlock(context);
```
Comment (msevestre) [inline src/MoBi.Core/Commands/RenameObjectBaseCommand.cs:42]: this feels weird. Why a RenameBuildingBlock method in a generic objectBaseCommand

File: src/MoBi.Presentation/DTO/RenameExpressionProfileDTO.cs [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+         {
+            _species = value;
+            OnPropertyChanged(() => Name);
+            OnPropertyChanged(() => Species);
+         }
+      }
+
+      private static class RenameExpressionProfileDTORules
+      {
+         public static IEnumerable<IBusinessRule> NonEmptyFieldRules()
```
Comment (msevestre) [inline src/MoBi.Presentation/DTO/RenameExpressionProfileDTO.cs:64]: change ths to be initalized once and not all the time.
Also return IReadOnly

File: src/MoBi.Presentation/DTO/RenameExpressionProfileDTO.cs [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+         {
+            yield return createNonEmptyRule(item => item.Species, Assets.AppConstants.Validation.SpeciesCannotBeEmpty);
+            yield return createNonEmptyRule(item => item.Category, Assets.AppConstants.Validation.CategoryCannotBeEmpty);
+            yield return createNonEmptyRule(item => item.MoleculeName, Assets.AppConstants.Validation.MoleculeNameCannotBeEmpty);
+         }
+
+         private static IBusinessRule createNonEmptyRule(Expression<Func<RenameExpressionProfileDTO, string>> propertyToCheck, string errorCaption)
```
Comment (msevestre) [inline src/MoBi.Presentation/DTO/RenameExpressionProfileDTO.cs:71]: change those method to use a get so that it's created only once and not every time

File: src/MoBi.Presentation/DTO/RenameExpressionProfileDTO.cs [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+         public static IBusinessRule UniqueName { get; } = CreateRule.For<RenameExpressionProfileDTO>()
+            .Property(item => item.Name)
+            .WithRule((dto, newName) => dto.isNameUnique(newName))
+            .WithError((dto, newName) => dto.nameAlreadyExistsError(newName));
+      }
+
+      private string nameAlreadyExistsError(string newName)
+      {
+         return Error.NameAlreadyExists(newName);
```
Comment (msevestre) [inline src/MoBi.Presentation/DTO/RenameExpressionProfileDTO.cs:87]: hum.. not sure about the gain of a method to call the same method somewhere else

File: src/MoBi.Presentation/Tasks/Edit/EditTaskFor.cs (hunk 2 above)
Comment (msevestre) [reply src/MoBi.Presentation/Tasks/Edit/EditTaskFor.cs:78]: what's the reason of doing this? You are adding dependencies in all derives class. There is probably a good reasons for this but I don't see it

File: src/MoBi.UI/Views/RenameExpressionProfileBuildingBlockView.cs [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+      {
+         _screenBinder.Bind(x => x.Category).To(tbCategory);
+         _screenBinder.Bind(x => x.MoleculeName).To(tbMoleculeName);
+         _screenBinder.Bind(x => x.Species).To(tbSpecies);
+         _screenBinder.Bind(x => x.Name).To(tbName);
+         _screenBinder.SavingMode = SavingMode.Always;
+
+         // We need to validate the text box when the text changes so that the Name property is validated
+         // even thought that text box is readonly. The Name is composed of each of the other text boxes content
+         tbCategory.TextChanged += (o,e) => OnEvent(() => validateTextBox(o));
```
Comment (msevestre) [inline src/MoBi.UI/Views/RenameExpressionProfileBuildingBlockView.cs:38]: I don't get this. why is this required? Isn't it what the property changes is all about?

**Review by msevestre — CHANGES_REQUESTED** [review]
See comments in PR

File: src/MoBi.Presentation/Tasks/Edit/EditTaskFor.cs (hunk 2 above)
Comment (rwmcintosh) [reply src/MoBi.Presentation/Tasks/Edit/EditTaskFor.cs:78]: It could be done another way.

What we had before was an ObjectBaseTask that did only this renaming in the same way for all objects. That dependency was part of InteractionTask which was part of InteractionTaskContext.

These two general dependencies could go in InteractionTaskContext so that not so many classes have direct dependency on these objects, only InteractionTaskContext would have to change.

I'll move it around and see how that fits.

File: src/MoBi.UI/Views/RenameExpressionProfileBuildingBlockView.cs (hunk 10 above)
Comment (rwmcintosh) [reply src/MoBi.UI/Views/RenameExpressionProfileBuildingBlockView.cs:38]: That's only updated when the field loses focus. To get this feature working the way I liked, I wanted to run validation on every keystroke.

File: tests/MoBi.Tests/Presentation/Tasks/InteractionTasksForRootApplicationBuilderSpecs.cs [hunk 11]
```diff
@@ -52,7 +50,7 @@ protected override void Because()
       [Observation]
       public void should_not_return_builder_name_as_forbidden()
       {
-         _result.ShouldOnlyContain(_builderName);
+         _result.ShouldNotContain(_builderName);
```
Comment (rwmcintosh) [inline tests/MoBi.Tests/Presentation/Tasks/InteractionTasksForRootApplicationBuilderSpecs.cs:53]: Trying to infer this test from the name and content. It feels like the original test was just using the wrong assertion.

File: src/MoBi.Presentation/Tasks/Edit/EditTaskFor.cs (hunk 2 above)
Comment (rwmcintosh) [reply src/MoBi.Presentation/Tasks/Edit/EditTaskFor.cs:75]: It's nicer this way. The dependencies are still the dependencies, but now these ones are bundled into InteractionTaskContext instead of in the constructors of all these EditTasks.

File: src/MoBi.UI/Views/RenameExpressionProfileBuildingBlockView.cs (hunk 10 above)
Comment (rwmcintosh) [reply src/MoBi.UI/Views/RenameExpressionProfileBuildingBlockView.cs:38]: I should say that the validation for that field already runs on each keystroke, but the underlying field is only updated on lost focus.

I want to update the underlying field and validate the 'Name' property on every keystroke on any one of these text boxes.

Comment (msevestre) [reply src/MoBi.UI/Views/RenameExpressionProfileBuildingBlockView.cs:38]: This should work out of the box. Not a big deal but if you have your DTO raising the right changed event...it happens automatically like for all other boxes.

**Review by msevestre — APPROVED** [review]
